moneydance: init at 2023.3_5064

lucasbergman opened the pull request:
## Description of changes

Moneydance is a proprietary, full featured financial application that runs on the JVM.

https://infinitekind.com/moneydance

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Steinhagen commented:
Gave it a try and works great! Everything else looks good to me.

ambroisie commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,58 @@
+    dpkg -x $src $out
+    mv $out/opt/Moneydance/lib $out/libexec
+    rm -rf $out/opt
+    makeWrapper ${jdk}/bin/java $out/bin/moneydance --add-flags "${flags}"
```
I don't know that we can rely on `escapeShellArgs` working correctly inside a double-quoted context.

Perhaps add multiple `--add-flags` parameters, one for each flag.

Something like this perhaps:
```nix
lib.flip lib.concatMapStrings flags (flag: "--add flags ${lib.escapeShellArg arg} ") 
```

lucasbergman replied:
Thanks @ambroisie. Experimentally, the thing I did _does_ work, but I actually like your version better for its clarity. What I did relies on rather specific and non-obvious behavior of makeWrapper.

While I've got you, can you help me figure out the error in https://gist.github.com/GrahamcOfBorg/5bd5b2c7f7b233bd83a32bab92efcf12 ? I assume my override is broken somehow, but I'm stumped. `nix-build -A moneydance` works locally. Does this mean my override trick is incompatible with `pkgs/by-name`?

ambroisie replied:
Not sure, this looks to be something specific to the JDK handling, perhaps go ask about it in Matrix.

ambroisie replied:
Looks like you could remove `headless = false`, as it's a work-around and is false by default.

```nix
# TODO(@sternenseemann): gtk3 fails to evaluate in pkgsCross.ghcjs.buildPackages
# which should be fixable, this is a no-rebuild workaround for GHC.
, headless ? stdenv.targetPlatform.isGhcjs
```

lucasbergman replied:
You're right, overriding with the `enableJavaFX` argument worked through OfBorg's eval, but the others didn't, even though local builds worked in all configurations. Hm, must be something specific to how the Java packages are built.

Steinhagen commented:
I believe that this commit should be split into 2. One for the maintainer part and one for the package itself:
1. maintainers: add lucasbergman
2. moneydance: init at 2023.3_5064

ambroisie commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,56 @@
+{ lib, stdenv, fetchurl, makeWrapper, dpkg, openjdk21, openjfx21, }:
+let
+  version = "2023.3_5064";
+  hash = "sha256-t1jBZJ2sL0YDx9915UWgpCzKnpNq1Qz5i7ygGXDI5dk=";
+in stdenv.mkDerivation rec {
+  pname = "moneydance";
+  inherit version;
+
+  src = fetchurl {
+    url = "https://infinitekind.com/stabledl/${version}/moneydance_linux_amd64.deb";
+    inherit hash;
+  };
```
Useless `let` block:

And let's use the `finalAttrs` idiom:
```suggestion
stdenv.mkDerivation (finalAttrs: {
  pname = "moneydance";
  version = "2023.3_5064";

  src = fetchurl {
    url = "https://infinitekind.com/stabledl/${finalAttrs.version}/moneydance_linux_amd64.deb";
    hash = "sha256-t1jBZJ2sL0YDx9915UWgpCzKnpNq1Qz5i7ygGXDI5dk=";
  };
```

ambroisie replied:
(Don't forget to close the parenthesis on the other side)

lucasbergman replied:
Huh, today I learned a thing!

lucasbergman committed:
maintainers: add lucasbergman

ambroisie commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,56 @@
+      "--add-modules"
+      "javafx.swing,javafx.controls,javafx.graphics"
+      "-classpath"
+      "'${placeholder "out"}/libexec/*'" # single-quotes to protect '*' from the shell
```
Is that necessary with `escapeShellArgs`?

lucasbergman replied:
Yes. The `escapeShellArg` call protects the `*` from `makeWrapper`, but it ends up unquoted in the generated script.

ambroisie replied:
Can we apply it twice then?

lucasbergman replied:
Yes, that would work, but IMHO this is starting to get a bit over-engineered? How do you like this new revision? It's quite explicit and contains (I believe) the minimum required shell cleverness.

lucasbergman committed:
moneydance: init at 2023.3_5064

Moneydance is a proprietary, full featured financial application that runs
on the JVM.

https://infinitekind.com/moneydance

fixes #277931

lucasbergman commented:
> I believe that this commit should be split into 2. One for the maintainer part and one for the package itself

Lovely. Done.

ambroisie commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,56 @@
+    runHook preInstall
+
+    mkdir -p $out $out/bin
+    dpkg -x $src $out
```
Rather than use `dpkg` and the `.deb` package, can you use the `moneydance-linux.tar.gz` tarball?

Or even better, package it from source?

lucasbergman replied:
Tarball it is! Packaging from source would be wonderful, but it's unfree / no source available, alas.

ambroisie commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,56 @@
+      "Moneydance is an easy to use and full-featured personal finance app that doesn't compromise your privacy";
+    sourceProvenance = [ lib.sourceTypes.binaryBytecode ];
+    license = lib.licenses.unfree;
+    platforms = lib.platforms.all;
```
The package mentions `amd64` (and implicitly linux), does it actually work on all platforms?

lucasbergman replied:
> Does it actually work on all platforms?

Yes. The package from the company is only platform specific because it ships a Java runtime. This package just strips all that out, and all that's left is JAR files and a driver shell script. Any platform with a Java runtime that supports JavaFX works.

ambroisie replied:
Then I would use `openjdk`'s platforms.

lucasbergman replied:
Oh, that's very clever. Done.

ambroisie commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,51 @@
+  pname = "moneydance";
+  version = "2023.3_5064";
+
+  src = fetchurl {
+    url = "https://infinitekind.com/stabledl/${finalAttrs.version}/moneydance-linux.tar.gz";
+    hash = "sha256-liU47bTKGBhDCnEQaKA0rUNPmBYH/zMJ7FOhmt5hQ14=";
+  };
```
Use `fetchzip` instead, this will take care of extracting the tarball.

lucasbergman replied:
Awesome

ambroisie commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,51 @@
+    hash = "sha256-liU47bTKGBhDCnEQaKA0rUNPmBYH/zMJ7FOhmt5hQ14=";
+  };
+
+  jdk = openjdk21.override { enableJavaFX = true; };
```
I'm not sure I quite like passing `jdk` as an argument to `mkDerivation`.

IMO it should be defined in a `let` block and `inherit`ed as `passthru.jdk` if you want to refer to it in other parts of the package definition.

lucasbergman replied:
Ah yes, that's good; I'm learning things about derivations today (I'm quite new to this). And that's not a totally theoretical benefit, since Moneydance supports plugins written in Java.

ambroisie commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,50 @@
+    hash = "sha256-jHr1V/gV1seenw2Q0/G405lTiabEYEsOS8p/XyByrtM=";
+  };
+
+  passthru = { inherit jdk; };
```
`passthru` is usally defined right above the `meta` block.

lucasbergman replied:
Oh, that makes sense. Done!

ambroisie commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,50 @@
+  meta = {
+    homepage = "https://infinitekind.com/moneydance";
+    description =
+      "Moneydance is an easy to use and full-featured personal finance app that doesn't compromise your privacy";
```
See [the documentation](https://nixos.org/manual/nixpkgs/unstable/#var-meta-description), you shouldn't repeat the name of package in the description.

```suggestion
      "An easy to use and full-featured personal finance app that doesn't compromise your privacy";
```

lucasbergman replied:
Very good; done. Thanks very much for the review.

ambroisie reviewed (approved):
Sorry, I missed one thing, otherwise LGTM.

I'll approve pending you apply that suggestion.

lucasbergman commented:
And @ambroisie and @Steinhagen, not strictly related to this PR: What do you think of [this followup change](https://github.com/lucasbergman/nixpkgs/commit/5b76f12684d39eacc50aa45df85ce8a721604f2c), and do you think it could go here or in its own PR (or neither)? I have a 4K screen, and Java UIs are unusable without that flag I give as an example in the commit message.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1421

ambroisie commented:
@lucasbergman I think it's fine to add it now

I would make the same suggestion about using something like `lib.concatMapStringSep " " (flag: "--add-flags ${lib.escapeShellArg arg}") jvmFlags;`

I would also rename the commit to `moneydance: add 'jvmFlags'`.

lucasbergman committed:
moneydance: add 'jvmFlags'

An example use case is that Java/Swing UIs, including Moneydance, are
often difficult to use on very high-resolution screens but can be fixed by
adding something like `-Dsun.java2d.uiScale=2` to JVM flags.

lucasbergman commented:
> @lucasbergman I think it's fine to add it now.

:+1:

> I would make the same suggestion about using something like `lib.concatMapStringSep " " (flag: "--add-flags ${lib.escapeShellArg arg}") jvmFlags;`.

Okay, you and I have circled around this a long time. Forgive me, but I think I've got a way of doing this that is decently explicit and readable but not too cute. I've verified experimentally that this works even if your list of flags contains values with spaces or cursed characters like `'`. Rationale:

1. IMO `--add-flags` is meant to loop over many flags and dump them on the command line; that's why its name is plural. 
2. There's no way to avoid double-escaping, and I _think_ this is simplest way to encode that in one spot.

This is a Nix string literal with a shell script that calls a shell script? function? to generate another shell script. It's a miracle that it's not even crazier.

So, what do you think? :rofl:

> I would also rename the commit to `moneydance: add 'jvmFlags'`.

Love it. Done.

ambroisie commented:
> This is a Nix string literal with a shell script that calls a shell script? function? to generate another shell script. It's a miracle that it's not even crazier.

Yes, I'm a bit unhappy at the situation due to the crazy quoting required, unfortunately I fear that it might be too late to change the implementation to make it more sane.

ambroisie reviewed (approved):
Yeah, I think this is the local minimum :-)

Thanks for your patience in this review. LGTM

lucasbergman commented:
Thanks for all your help, folks. Is there something I should do in order to get merged?

ambroisie commented:
@lucasbergman you can ask for (merge-bit-enabled) reviewers to take a look on discourse (the "PRs already reviewed" thread).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3374

SuperSandro2000 commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,58 @@
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ jdk openjfx21 ];
+
+  dontConfigure = true;
+  dontUnpack = true;
+  dontBuild = true;
+  dontFixup = true;
```
Break those things? If not please leave them enabled. Especially fixup phase

lucasbergman replied:
For this particular kind of package (pure Java), those phases don't do anything in practice, but leaving them doesn't break anything and is only a tiny bit slower. Reverted.

SuperSandro2000 commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,58 @@
+
+  meta = {
+    homepage = "https://infinitekind.com/moneydance";
+    description =
+      "An easy to use and full-featured personal finance app that doesn't compromise your privacy";
```
```suggestion
    description = "An easy to use and full-featured personal finance app that doesn't compromise your privacy";
```

lucasbergman replied:
Oops, good catch. Done.

SuperSandro2000 commented on pkgs/by-name/mo/moneydance/package.nix:
```diff
@@ -0,0 +1,58 @@
+      "javafx.swing,javafx.controls,javafx.graphics"
+      "-classpath"
+      "${placeholder "out"}/libexec/*"
+    ] ++ jvmFlags ++ [ "Moneydance" ];
```
Lets take a step back: why do we need the jvmFlags option? Is that very commonly changed?

lucasbergman replied:
To be totally honest, I can't be sure of how widespread that is. My personal use case is setting `-Dsun.java2d.uiScale=2` since Java Swing UIs are unusable on very high-resolution displays without it. Thanks for the review. I'm super-new to NixOS, so I'm happy to follow whatever customs you prefer.

lucasbergman replied:
Friendly ping for @SuperSandro2000
